File: 01-aop/with-aop/src/main/java/com/example/withaop/repository/UserRepository.java
package com.example.withaop.repository;

import com.example.withaop.model.User;
import org.springframework.stereotype.Repository;

import java.util.HashMap;
import java.util.Map;

@Repository
public class UserRepository {
    private final Map<Long, User> users = new HashMap<>();
    private Long sequence = 0L;

    public UserRepository() {
        // 초기 데이터
        save(new User(null, "John Doe", "john@example.com"));
        save(new User(null, "Jane Smith", "jane@example.com"));
    }

    public User findById(Long id) {
        User user = users.get(id);
        if (user == null) {
            throw new RuntimeException("사용자를 찾을 수 없습니다: " + id);
        }
        return user;
    }

    public User save(User user) {
        if (user.getId() == null) {
            user.setId(++sequence);
        }
        users.put(user.getId(), user);
        return user;
    }

    public void deleteById(Long id) {
        if (!users.containsKey(id)) {
            throw new RuntimeException("사용자를 찾을 수 없습니다: " + id);
        }
        users.remove(id);
    }
}

File: 02-di-ioc/1-manual-di/src/main/kotlin/com/example/manual/repository/UserRepository.kt
package com.example.manual.repository

import com.example.manual.model.User

/**
 * 사용자 저장소 인터페이스
 */
interface UserRepository {
    fun save(user: User): User
    fun findById(id: Long): User?
    fun findAll(): List<User>
}

/**
 * 사용자 저장소 구현체 (메모리 기반)
 *
 * 문제점: UserService가 이 구체적인 구현체를 직접 생성해야 함
 */
class UserRepositoryImpl : UserRepository {
    private val users = mutableMapOf<Long, User>()
    private var sequence = 0L

    override fun save(user: User): User {
        val id = user.id ?: ++sequence
        val saved = user.copy(id = id)
        users[id] = saved
        println("[Repository] 사용자 저장: ${saved.name} (ID: $id)")
        return saved
    }

    override fun findById(id: Long): User? {
        return users[id]
    }

    override fun findAll(): List<User> {
        return users.values.toList()
    }
}
